test/hotspot/jtreg/compiler/lib/template_framework/Renderer.java

```diff
@@ -243,82 +224,181 @@ private static String formatDouble(Double d) {
     }
 
     private void renderTemplateToken(TemplateToken templateToken) {
+        // We need a TemplateFrame in all cases, this ensures that the outermost scope of the template
+        // is not transparent for hashtags and setFuelCost, and also that the id of the template is
+        // unique.
         TemplateFrame templateFrame = TemplateFrame.make(currentTemplateFrame, nextTemplateFrameId++);
         currentTemplateFrame = templateFrame;
 
         templateToken.visitArguments((name, value) -> addHashtagReplacement(name, format(value)));
-        TemplateBody body = templateToken.instantiate();
-        renderTokenList(body.tokens());
+
+        // If the ScopeToken is transparent to Names, then the Template is transparent to names.
+        ScopeToken st = templateToken.instantiate();
+        renderScopeToken(st, () -> {});
 
         if (currentTemplateFrame != templateFrame) {
             throw new RuntimeException("Internal error: TemplateFrame mismatch!");
         }
         currentTemplateFrame = currentTemplateFrame.parent;
     }
 
+    private void renderScopeToken(ScopeToken st, Runnable preamble) {
+        if (!(st instanceof ScopeTokenImpl sti)) {
```

Comment:
Wasn't aware of that but seems quite convenient: The IDE suggests to use a record pattern:

```
if (!(st instanceof ScopeTokenImpl(List<Token> tokens, boolean nestedNamesAreLocal,
                                   boolean nestedHashtagsAreLocal, boolean nestedSetFuelCostAreLocal
))) {
```
Then you can directly access the fields below:
```
sti.nestedNamesAreLocal() -> nestedNamesAreLocal
```
etc.